cpp/src/parquet/page_index_test.cc

```diff
@@ -0,0 +1,260 @@
+                                             ByteArray{max_value}};
+
+  TestReadTypedColumnIndex<ByteArrayType>(
+      "alltypes_tiny_pages.parquet", column_id, num_pages, boundary_order, page_indices,
+      null_pages, min_values, max_values, has_null_counts, null_counts);
+}
+
+TEST(PageIndex, ReadBoolColumnIndex) {
+  const int column_id = 1;
+  const size_t num_pages = 82;
+  const BoundaryOrder::type boundary_order = BoundaryOrder::Ascending;
+  const std::vector<size_t> page_indices = {0, 16, 64};
+  const std::vector<bool> null_pages = {false, false, false};
+  const bool has_null_counts = true;
+  const std::vector<int64_t> null_counts = {0, 0, 0};
+  const std::vector<bool> min_values = {false, false, false};
+  const std::vector<bool> max_values = {true, true, true};
+
+  TestReadTypedColumnIndex<BooleanType>(
+      "alltypes_tiny_pages.parquet", column_id, num_pages, boundary_order, page_indices,
+      null_pages, min_values, max_values, has_null_counts, null_counts);
+}
+
+TEST(PageIndex, FixedLengthByteArrayColumnIndex) {
+  auto to_flba = [](const char* ptr) {
+    return FLBA{reinterpret_cast<const uint8_t*>(ptr)};
+  };
+
+  const int column_id = 0;
+  const size_t num_pages = 10;
+  const BoundaryOrder::type boundary_order = BoundaryOrder::Descending;
+  const std::vector<size_t> page_indices = {0, 4, 8};
+  const std::vector<bool> null_pages = {false, false, false};
+  const bool has_null_counts = true;
+  const std::vector<int64_t> null_counts = {9, 13, 9};
+  const std::vector<const char*> min_literals = {"\x00\x00\x03\x85", "\x00\x00\x01\xF5",
+                                                 "\x00\x00\x00\x65"};
+  const std::vector<const char*> max_literals = {"\x00\x00\x03\xE8", "\x00\x00\x02\x58",
+                                                 "\x00\x00\x00\xC8"};
+  const std::vector<FLBA> min_values = {
+      to_flba(min_literals[0]), to_flba(min_literals[1]), to_flba(min_literals[2])};
+  const std::vector<FLBA> max_values = {
+      to_flba(max_literals[0]), to_flba(max_literals[1]), to_flba(max_literals[2])};
+
+  TestReadTypedColumnIndex<FLBAType>(
+      "fixed_length_byte_array.parquet", column_id, num_pages, boundary_order,
+      page_indices, null_pages, min_values, max_values, has_null_counts, null_counts);
+}
+
+TEST(PageIndex, ReadColumnIndexWithNullPage) {
+  const int column_id = 0;
+  const size_t num_pages = 2;
+  const BoundaryOrder::type boundary_order = BoundaryOrder::Ascending;
+  const std::vector<size_t> page_indices = {0, 1};
+  const std::vector<bool> null_pages = {true, true};
+  // It seems that the null_counts are malformed.
+  const bool has_null_counts = true;
+  const std::vector<int64_t> null_counts = {-1, -1};
+  const std::vector<int32_t> min_values = {};
+  const std::vector<int32_t> max_values = {};
+
+  TestReadTypedColumnIndex<Int32Type>(
+      "datapage_v1-uncompressed-checksum.parquet", column_id, num_pages, boundary_order,
```

Comment:
If you read this file, you'll see that all data is actually non-null. So the column index in this file may be bogus?
(or you're reading it wrong?)

Reply:
I think the page index is corrupted. The parquet-cli has provided same reading:

```
➜ parquet-cli column-index datapage_v1-uncompressed-checksum.parquet
row-group 0:
column index for column a:
Boundary order: ASCENDING
                      null count  min                                       max
page-0                        -1  <none>                                    <none>
page-1                        -1  <none>                                    <none>

offset index for column a:
                          offset   compressed size       first row index
page-0                         4             10268                     0
page-1                     10272             10268                  2560

column index for column b:
Boundary order: ASCENDING
                      null count  min                                       max
page-0                        -1  <none>                                    <none>
page-1                        -1  <none>                                    <none>

offset index for column b:
                          offset   compressed size       first row index
page-0                     20540             10268                     0
page-1                     30808             10268                  2560
```